File: fixtures/sast/bad-sample.py
# ============================================================================
# fixtures/sast/bad-sample.py — SAST 坏样本（#146 · v0.5.22）
# ----------------------------------------------------------------------------
# 故意携带 5 类危险模式（测试专用语料，密钥均为假样例形状，非真实凭据）：
#   ① 硬编码密钥（sk-ant- / ghp_ / AKIA 三形态 —— 内置密钥规则面）
#   ② eval 动态执行（ruff S307 + 内置注入规则面）
#   ③ SQL 字符串拼接（ruff S608 + 内置 SQL 规则面）
#   ④ shell 拼接（ruff S605 + 内置 shell 规则面）
#   ⑤ 弱随机做 token（ruff S311 + 内置弱随机规则面）
# 双车道对拍：tests/sast.test.ts 用本文件跑 ruff 车道（--select S）与内置
# 降级车道（engine:"builtin"）双对拍 —— 两车道各有产出且规则面互补
# （ruff 拿 S 码诊断；内置拿密钥形态 —— ruff 只按变量名认密钥，内置认值形状）。
# ============================================================================

import os
import random

# ① 硬编码密钥（三形态假样例）
API_KEY = "sk-ant-abcdefghijklmnopqrstuvwxyz123456"
GH_TOKEN = "ghp_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijkl"
AWS_KEY = "AKIAIOSFODNN7EXAMPLE"


# ② eval 动态执行（注入面）
def run_user_code(code):
    return eval(code)


# ③ SQL 字符串拼接（注入面）
def load_user(user_id):
    sql = "SELECT * FROM users WHERE id = " + user_id
    return sql


# ④ shell 拼接（注入面）
def dump_file(fname):
    os.system("cat " + fname)


# ⑤ 弱随机做 token（密码学用途）
def make_token():
    return random.random()

File: fixtures/sast/bad-sample.ts
// ============================================================================
// fixtures/sast/bad-sample.ts — SAST 坏样本 · TS 侧（#146 · v0.5.22）
// ----------------------------------------------------------------------------
// 与 bad-sample.py 同族的 5 类危险模式（测试专用语料，密钥为假样例形状）：
// 内置 TS 规则面（eval/new Function / SQL 模板字面量拼接 / exec 模板拼接 /
// Math.random 做 token / 硬编码密钥）。ruff 车道只覆盖 Python —— TS 侧由
// 内置降级规则引擎兜底（这正是降级链的「永远有产出」半面）。
// ============================================================================

// ① 硬编码密钥（三形态假样例）
export const API_KEY = "sk-ant-abcdefghijklmnopqrstuvwxyz123456";
export const GH_TOKEN = "ghp_ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijkl";
export const AWS_KEY = "AKIAIOSFODNN7EXAMPLE";

// ② eval / new Function 动态执行（注入面）
export function runUserCode(code: string): unknown {
  return eval(code);
}
export function buildFn(body: string): Function {
  return new Function("x", body);
}

// ③ SQL 模板字面量拼接（注入面）
export function loadUser(userId: string): string {
  return `SELECT * FROM users WHERE id = ${userId}`;
}

// ④ exec 模板拼接（注入面）
import { execSync } from "node:child_process";
export function dumpFile(fname: string): string {
  return execSync(`cat ${fname}`).toString();
}

// ⑤ 弱随机做 token（密码学用途）
export function makeToken(): string {
  return Math.random().toString(36);
}
